generate-zephyr-dts.py

```diff
@@ -1,179 +1,136 @@
+def peripheral_handler(name, parm, csr):
+    dtsi  = dts_reg(" ".join([
+                hex(csr['csr_bases'][name]),
+                hex(parm['size'])]))
+    try:
+        dtsi += dts_intr(name, csr)
+    except KeyError as e:
+        print('  dtsi key', e, 'not found, no interrupt override')
+    return dtsi
 
-    if 'constants' in peripheral:
-        if 'interrupt' in peripheral['constants']:
-            irq_no = int(peripheral['constants']['interrupt'], 0)
-            result += """
-    interrupts = <{} 0>;
-""".format(hex(irq_no))
 
-    result += """
-};
-"""
-    return result
-
-
-def i2c_handler(peripheral, **kw):
-    result = """
-&{} {{
-    reg = <{} {} {} {}>;
-}};
-""".format(kw['reference'],
-           hex(peripheral['address']),
-           hex(kw['size']),
-           hex(peripheral['address'] + kw['size']),
-           hex(kw['size']))
-
-    return result
-
-
-def peripheral_handler(peripheral, **kw):
-    result = """
-&{} {{
-""".format(kw['reference'])
-
-    result += """
-    reg = <{} {}>;
-""".format(hex(peripheral['address']),
-           hex(kw['size']))
-
-    if 'constants' in peripheral:
-        if 'interrupt' in peripheral['constants']:
-            irq_no = int(peripheral['constants']['interrupt'], 0)
-            result += """
-    interrupts = <{} 0>;
-""".format(hex(irq_no))
-
-    result += """
-};
-"""
-    return result
-
-
-peripheral_handlers = {
+overlay_handlers = {
     'uart': {
         'handler': peripheral_handler,
-        'reference': 'uart0',
-        'size': 0x18,
+        'alias': 'uart0',
+        'size': 0x20,
         'config_entry': 'UART_LITEUART'
     },
     'timer0': {
         'handler': peripheral_handler,
-        'reference': 'timer0',
         'size': 0x40,
         'config_entry': 'LITEX_TIMER'
     },
     'ethmac': {
         'handler': ethmac_handler,
-        'reference': 'eth0',
-        'size': 0x6c,
-        'buffer': lambda: configuration.mem_regions['ethmac'],
+        'alias': 'eth0',
+        'size': 0x80,
         'config_entry': 'ETH_LITEETH'
     },
+    'spiflash': {
+        'handler': peripheral_handler,
+        'alias': 'spi0',
+        'size': 12,
+        'config_entry': 'SPI_LITESPI'
+    },
     'i2c0' : {
         'handler': i2c_handler,
-        'reference': 'i2c0',
         'size': 0x4,
         'config_entry': 'I2C_LITEX'
-    }
-}
-
-
-mem_region_handler = {
+    },
     'main_ram': {
         'handler': ram_handler,
-    }
+        'alias': 'ram0',
+    },
+    # disabled
+    'dna0': {
+        'handler': disabled_handler,
+    },
 }
 
 
-def generate_dts():
-    result = ""
+def generate_dts_config(csr):
+    dts = cnf = ''
 
-    for name, peripheral in configuration.peripherals.items():
-        if name not in peripheral_handlers:
-            print('Skipping unsupported peripheral `{}` at {}'
-                  .format(name, hex(peripheral['address'])))
-            continue
+    for name, parm in overlay_handlers.items():
```

Comment:
By iterating over `overlay_handler` instead of `configuration.peripherals` you miss the warnings about unsupported (yet available in the platform) peripherals.
I think this might be important when debugging problems.

Reply:
Hi @mateusz-holenko,

Thanks you for the review!

Yes I can see your point, the shared code between the DTS and Renode configuration.

Currently there are 6 parsers or interpretations made:
`LiteX to CSV  ---  CSV to lx.configutation`
`LiteX to JSON  ---  JSON to lx.configutation`
`lx.configuration to renode`
`lx.configutation to DTS`

I would argue to at least ditch the `CSV` format in favor of `JSON` as it is better at preserving structure and type. So `JSON` seems more future proof, and we can ditch `CSV` at zero cost, nobody gets hurts as fat as i can tell..
So then we have to maintain two parsers less:

`LiteX to JSON  ---  JSON to lx.configutation`
`lx.configuration to renode`
`lx.configutation to DTS`

Similarl since the `CSV` is a subset of `JSON` and there are no complex routines that can be shared, for now. The benefits of having a separate configuration class just for abstracting the JSON and CSV inputs is not that usefull. So you could further reduce to:

`LiteX to JSON`
`JSON to renode`
`JSON to DTS`

We end up with two parsers closely coupled to the JSON output format of LITEX.

Reply:
> By iterating over `overlay_handler` instead of `configuration.peripherals` you miss the warnings about unsupported (yet available in the platform) peripherals.
> I think this might be important when debugging problems.

I agree this should be fixed!

Reply:
> we can ditch CSV at zero cost, nobody gets hurts as far as i can tell

Do you mean that we can parse CSV files using the json parser?  litex-buildenv uses this script to generate Zephyr overlay from CSV: https://github.com/timvideos/litex-buildenv/blob/master/scripts/build-zephyr.sh#L84.

Reply:
Additionally https://github.com/google/CFU-Playground also uses the csv format for generating Renode scripts at the moment.

Reply:
Nice links, both are interesting projects. 
The main drawback for JSON is that the addresses are in decimal instead of hex, so it is not as human readable as CSV.